internal/provisioner/objects/rbac/util/util.go

```diff
@@ -0,0 +1,86 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	gatewayv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+const (
+	ContourV1GroupName = "projectcontour.io"
+)
+
+var (
+	createGetUpdate = []string{"create", "get", "update"}
+	getListWatch    = []string{"get", "list", "watch"}
+	update          = []string{"update"}
+)
+
+// PolicyRuleFor returns PolicyRule object with provided apiGroup, verbs and resources
+func PolicyRuleFor(apiGroup string, verbs []string, resources ...string) rbacv1.PolicyRule {
+	return rbacv1.PolicyRule{
+		Verbs:     verbs,
+		APIGroups: []string{apiGroup},
+		Resources: resources,
+	}
+}
+
+// NamespacedResourcePolicyRules returns a set of policy rules for resources that are
+// namespaced-scoped
+func NamespacedResourcePolicyRules() []rbacv1.PolicyRule {
+	return []rbacv1.PolicyRule{
+		// Core Contour-watched resources.
+		PolicyRuleFor(corev1.GroupName, getListWatch, "secrets", "endpoints", "services"),
+
+		// Discovery Contour-watched resources.
+		PolicyRuleFor(discoveryv1.GroupName, getListWatch, "endpointslices"),
+
+		// Gateway API resources.
+		// Note, ReferenceGrant does not currently have a .status field so it's omitted from the status rule.
+		PolicyRuleFor(gatewayv1alpha2.GroupName, getListWatch, "gateways", "httproutes", "tlsroutes", "grpcroutes", "tcproutes", "referencegrants"),
+		PolicyRuleFor(gatewayv1alpha2.GroupName, update, "gateways/status", "httproutes/status", "tlsroutes/status", "grpcroutes/status", "tcproutes/status"),
+
+		// Ingress resources.
+		PolicyRuleFor(networkingv1.GroupName, getListWatch, "ingresses"),
+		PolicyRuleFor(networkingv1.GroupName, createGetUpdate, "ingresses/status"),
+
+		// Contour CRDs.
+		PolicyRuleFor(ContourV1GroupName, getListWatch, "httpproxies", "tlscertificatedelegations", "extensionservices", "contourconfigurations"),
+		PolicyRuleFor(ContourV1GroupName, createGetUpdate, "httpproxies/status", "extensionservices/status", "contourconfigurations/status"),
+	}
+}
+
+// ClusterScopedResourcePolicyRules returns a set of policy rules for
+// cluster-scoped resources.
+func ClusterScopedResourcePolicyRules() []rbacv1.PolicyRule {
+	return []rbacv1.PolicyRule{
+		// GatewayClass.
+		PolicyRuleFor(gatewayv1alpha2.GroupName, getListWatch, "gatewayclasses"),
+		PolicyRuleFor(gatewayv1alpha2.GroupName, update, "gatewayclasses/status"),
+
+		// Namespaces
+		PolicyRuleFor(corev1.GroupName, getListWatch, "namespaces"),
+	}
+}
+
+// ContourResourceName returns names for roles/rolebindings for contour resources
+func ContourResourceName(name string) string {
+	return fmt.Sprintf("contour-resource-%s", name)
+}
```

Comment:
Let's move this into `names.go` where the rest of the resource names are defined. We can add a new field (something like `ResourceNamespaceRole`) under `RBACNames` to store this. Also, since these roles/rolebindings are potentially being created outside the namespace where the Gateway is created, let's ensure the Role/RoleBinding name includes both the namespace and name of the Gateway. So, for example, if the Gateway is named `gateway-1` inside the `projectcontour` namespace, then these roles/rolebindings can be named `contour-resources-projectcontour-gateway-1`. That's most consistent with existing naming conventions, I think.